server/src/main/java/com/saralein/server/controller/PartialContentController.java

```diff
@@ -0,0 +1,134 @@
+                .status(416)
+                .addHeader("Content-Range", "*/" + fileLength)
+                .build();
+    }
+
+    private Response buildValidRangeResponse(Request request, int fileLength, int start, int end) throws IOException {
+        Path file = fileHelper.createAbsolutePath(request.getUri());
+        byte[] body = fileIO.readPartialFile(file, start, end);
+
+        return new Response.Builder()
+                .status(206)
+                .addHeader("Content-Type", fileHelper.determineMimeType(request.getUri()))
+                .addHeader("Content-Range", formatContentRange(start, end, fileLength))
+                .body(body)
+                .build();
+    }
+
+    private boolean requestInRange(int fileLength, int start, int end) {
+        return (start >= 0) && (start < end) && (end < fileLength);
+    }
+
+    private int[] parseRange(String range, int fileLength) {
+        if (specifiesRangeFromEnd(range)) {
+            return parseRangeFromEnd(range, fileLength);
+        }
+
+        if (specifiesRangeFromStart(range)) {
+            return parseRangeFromStart(range, fileLength);
+        }
+
+        return parseFullRange(range);
+    }
+
+    private boolean specifiesFullRange(String range) {
+        return range.matches("bytes=\\d+-\\d+");
+    }
+
+    private boolean specifiesRangeFromStart(String range) {
+        return range.matches("bytes=\\d+-");
+    }
+
+    private boolean specifiesRangeFromEnd(String range) {
+        return range.matches("bytes=-\\d+");
+    }
+
+    private boolean validRangeFormat(String range) {
+        return specifiesFullRange(range)
+                || specifiesRangeFromStart(range)
+                || specifiesRangeFromEnd(range);
+    }
+
+    private String removeRangeUnit(String range) {
+        return range.replace("bytes=", "");
+    }
+
+    private int[] parseRangeFromEnd(String range, int fileLength) {
+        String trimmedRange = removeRangeUnit(range);
+        int value = Integer.parseInt(trimmedRange.substring(1));
+        return new int[]{fileLength - value, fileLength - 1};
+    }
+
+    private int[] parseRangeFromStart(String range, int fileLength) {
+        String trimmedRange = removeRangeUnit(range);
+        int value = Integer.parseInt(trimmedRange.substring(0, trimmedRange.length() - 1));
+        return new int[]{value, fileLength - 1};
+    }
+
+    private int[] parseFullRange(String range) {
+        String trimmedRange = removeRangeUnit(range);
+        String[] splitRange = trimmedRange.split("-");
+        return new int[]{Integer.parseInt(splitRange[0]), Integer.parseInt(splitRange[1])};
+    }
+
+    private int rangeStart(int[] range) {
+        return range[0];
+    }
+
+    private int rangeEnd(int[] range) {
+        return range[1];
+    }
+
+    private String formatContentRange(int start, int end, int fileLength) {
+        return String.format("bytes %d-%d/%d", start, end, fileLength);
+    }
+}
```

Comment:
There are a handful of methods in this class that share a prefix ("parse", "range", "specifies").

Reply:
Tried breaking this up into smaller classes based on this.